adserver/tasks.py

```diff
@@ -143,3 +148,48 @@ def daily_update_impressions(day=None):
             AdImpression.objects.filter(pk=impression.pk).update(
                 **{impression_type: values["total"]}
             )
+
+
+@app.task()
+def daily_update_keywords(day=None):
+    """
+    Update the KeywordImpression index each day.
+
+    :arg day: An optional datetime object representing a day
+    """
+    start_date, end_date = _get_day(day)
+
+    log.info("Updating KeywordImpression for %s-%s", start_date, end_date)
+
+    for impression_type in IMPRESSION_TYPES:
+        queryset = _default_filters(impression_type, start_date, end_date)
+
+        for values in (
+            queryset.values("publisher", "advertisement", "keywords")
+            .annotate(total=Count("keywords"))
+            .filter(total__gt=0)
+            .order_by("-keywords")
+        ):
```

Comment:
This gets the unique publisher/advertisement/keywords combination for a day. This is going to be enormously expensive especially for publishers like Read the Docs. Did you try this in production to see how long it takes and how many results it yields? I am worried about running this every 30 minutes.

Reply:
It runs hourly, but we could also scale that back if we're worried about it. I haven't tested it against RTD.

Reply:
I tested it with today's date in prod and it ran in ~10 seconds.

Reply:
Removing the inline queries makes it really fast 👍